From: "B. Stack" <bgstack15@gmail.com>
Date: Sat, 15 Mar 2025 09:03:12 +0100
Subject: Disable some functionality that depends on wx3.3 which is not yet in
 Debian. Unfortunately, this includes dark mode

Last-Modified: 2025-02-20
Last-Version: 14.2
Forwarded: not-needed


Required since 14.0
---
 FreeFileSync/Source/ui/cfg_grid.cpp           |  4 ++--
 FreeFileSync/Source/ui/file_grid.cpp          | 16 +++++++--------
 FreeFileSync/Source/ui/gui_status_handler.cpp | 11 +++++++++--
 FreeFileSync/Source/ui/log_panel.cpp          |  4 ++--
 FreeFileSync/Source/ui/main_dlg.cpp           | 15 ++++++++++++--
 FreeFileSync/Source/ui/rename_dlg.cpp         | 16 ++++++++-------
 FreeFileSync/Source/ui/small_dlgs.cpp         |  2 +-
 FreeFileSync/Source/ui/tree_grid.cpp          |  4 ++--
 wx+/darkmode.cpp                              | 11 ++++++-----
 wx+/darkmode.h                                | 28 ++++++++++++++++++++++++++-
 wx+/grid.cpp                                  | 13 +++++++------
 wx+/grid.h                                    |  4 ++--
 12 files changed, 88 insertions(+), 40 deletions(-)

diff --git a/FreeFileSync/Source/ui/cfg_grid.cpp b/FreeFileSync/Source/ui/cfg_grid.cpp
index 0cd3a2f..e1453b7 100644
--- a/FreeFileSync/Source/ui/cfg_grid.cpp
+++ b/FreeFileSync/Source/ui/cfg_grid.cpp
@@ -531,7 +531,7 @@ private:
             }
     }
 
-    int getBestSize(const wxReadOnlyDC& dc, size_t row, ColumnType colType) override
+    int getBestSize(wxDC& dc, size_t row, ColumnType colType) override
     {
         // -> synchronize renderCell() <-> getBestSize()
 
@@ -550,7 +550,7 @@ private:
         return 0;
     }
 
-    HoverArea getMouseHover(const wxReadOnlyDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) override
+    HoverArea getMouseHover(wxDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) override
     {
         if (const ConfigView::Details* item = cfgView_.getItem(row))
         {
diff --git a/FreeFileSync/Source/ui/file_grid.cpp b/FreeFileSync/Source/ui/file_grid.cpp
index 7062b7a..32713e9 100644
--- a/FreeFileSync/Source/ui/file_grid.cpp
+++ b/FreeFileSync/Source/ui/file_grid.cpp
@@ -341,7 +341,7 @@ public:
 
     const FileSystemObject* getFsObject(size_t row) const { return getDataView().getFsObject(row); }
 
-    const wxSize& getTextExtentBuffered(const wxReadOnlyDC& dc, const std::wstring_view& text)
+    const wxSize& getTextExtentBuffered(wxDC& dc, const std::wstring_view& text)
     {
         auto& compExtentsBuf = sharedComp_.ref().compExtentsBuf_;
         //- only used for parent path names and file names on view => should not grow "too big"
@@ -357,7 +357,7 @@ public:
 
     //- trim while leaving path components intact
     //- *always* returns at least one component, even if > maxWidth
-    size_t getPathTrimmedSize(const wxReadOnlyDC& dc, const std::wstring_view& itemPath, int maxWidth)
+    size_t getPathTrimmedSize(wxDC& dc, const std::wstring_view& itemPath, int maxWidth)
     {
         if (itemPath.size() <= 1)
             return itemPath.size();
@@ -586,7 +586,7 @@ private:
     }
 
 
-    int getGroupItemNamesWidth(const wxReadOnlyDC& dc, const FileView::PathDrawInfo& pdi)
+    int getGroupItemNamesWidth(wxDC& dc, const FileView::PathDrawInfo& pdi)
     {
         //FileView::updateView() called? => invalidates group item render buffer
         if (pdi.viewUpdateId != viewUpdateIdLast_)
@@ -640,7 +640,7 @@ private:
         int groupParentWidth;
         int groupNameWidth;
     };
-    GroupRowLayout getGroupRowLayout(const wxReadOnlyDC& dc, size_t row, const FileView::PathDrawInfo& pdi, int maxWidth)
+    GroupRowLayout getGroupRowLayout(wxDC& dc, size_t row, const FileView::PathDrawInfo& pdi, int maxWidth)
     {
         assert(pdi.fsObj);
 
@@ -1179,7 +1179,7 @@ private:
     }
 
 
-    HoverArea getMouseHover(const wxReadOnlyDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) override
+    HoverArea getMouseHover(wxDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) override
     {
         if (static_cast<ColumnTypeRim>(colType) == ColumnTypeRim::path)
             if (const FileView::PathDrawInfo pdi = getDataView().getDrawInfo(row);
@@ -1203,7 +1203,7 @@ private:
     }
 
 
-    int getBestSize(const wxReadOnlyDC& dc, size_t row, ColumnType colType) override
+    int getBestSize(wxDC& dc, size_t row, ColumnType colType) override
     {
         if (static_cast<ColumnTypeRim>(colType) == ColumnTypeRim::path)
         {
@@ -1235,7 +1235,7 @@ private:
         }
         else
         {
-            const wxReadOnlyDC& infoDc = dc;
+            const wxDC& infoDc = dc;
             const std::wstring cellValue = getValue(row, colType);
             return gapSize_ + infoDc.GetTextExtent(cellValue).GetWidth() + gapSize_;
         }
@@ -1598,7 +1598,7 @@ private:
         }
     }
 
-    HoverArea getMouseHover(const wxReadOnlyDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) override
+    HoverArea getMouseHover(wxDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) override
     {
         if (const FileSystemObject* const fsObj = getFsObject(row))
             switch (static_cast<ColumnTypeCenter>(colType))
diff --git a/FreeFileSync/Source/ui/gui_status_handler.cpp b/FreeFileSync/Source/ui/gui_status_handler.cpp
index f3a3df3..c3f6000 100644
--- a/FreeFileSync/Source/ui/gui_status_handler.cpp
+++ b/FreeFileSync/Source/ui/gui_status_handler.cpp
@@ -70,27 +70,34 @@ void StatusHandlerTemporaryPanel::showStatsPanel()
                 //case wxAUI_DOCK_CENTRE:
         }
 
+        wxAuiPaneInfoArray& paneArray = mainDlg_.auiMgr_.GetAllPanes();
         const bool statusRowTaken = [&]
         {
-            for (wxAuiPaneInfo& paneInfo : mainDlg_.auiMgr_.GetAllPanes())
+            for (size_t i = 0; i < paneArray.size(); ++i)
+            {
+                const wxAuiPaneInfo& paneInfo = paneArray[i];
                 //doesn't matter if paneInfo.IsShown() or not! => move down in either case!
                 if (&paneInfo != &statusPanel &&
                     paneInfo.dock_layer     == statusPanel.dock_layer &&
                     paneInfo.dock_direction == statusPanel.dock_direction &&
                     paneInfo.dock_row       == statusPanel.dock_row)
                     return true;
+            }
 
             return false;
         }();
 
         //move all rows that are in the way one step further
         if (statusRowTaken)
-            for (wxAuiPaneInfo& paneInfo : mainDlg_.auiMgr_.GetAllPanes())
+            for (size_t i = 0; i < paneArray.size(); ++i)
+            {
+                wxAuiPaneInfo& paneInfo = paneArray[i];
                 if (&paneInfo != &statusPanel &&
                     paneInfo.dock_layer     == statusPanel.dock_layer &&
                     paneInfo.dock_direction == statusPanel.dock_direction &&
                     paneInfo.dock_row       >= statusPanel.dock_row)
                     ++paneInfo.dock_row;
+            }
         //------------------------------------------------------------------
 
         statusPanel.Show();
diff --git a/FreeFileSync/Source/ui/log_panel.cpp b/FreeFileSync/Source/ui/log_panel.cpp
index 494264b..c7a0eb5 100644
--- a/FreeFileSync/Source/ui/log_panel.cpp
+++ b/FreeFileSync/Source/ui/log_panel.cpp
@@ -245,7 +245,7 @@ public:
             }
     }
 
-    int getBestSize(const wxReadOnlyDC& dc, size_t row, ColumnType colType) override
+    int getBestSize(wxDC& dc, size_t row, ColumnType colType) override
     {
         // -> synchronize renderCell() <-> getBestSize()
 
@@ -266,7 +266,7 @@ public:
 
     static int getColumnTimeDefaultWidth(Grid& grid)
     {
-        wxInfoDC dc(&grid.getMainWin());
+        wxClientDC dc(&grid.getMainWin());
         dc.SetFont(grid.getMainWin().GetFont());
         return 2 * getColumnGapLeft() + dc.GetTextExtent(utfTo<wxString>(formatTime(formatTimeTag))).GetWidth();
     }
diff --git a/FreeFileSync/Source/ui/main_dlg.cpp b/FreeFileSync/Source/ui/main_dlg.cpp
index 1c57523..7963488 100644
--- a/FreeFileSync/Source/ui/main_dlg.cpp
+++ b/FreeFileSync/Source/ui/main_dlg.cpp
@@ -1277,8 +1277,9 @@ void MainDialog::setGlobalCfgOnInit(const GlobalConfig& globalCfg)
 
     //work around wxAuiManager::LoadPerspective overwriting pane captions with old values (might be different language!)
     std::vector<std::pair<wxAuiPaneInfo*, wxString>> paneCaptions;
-    for (wxAuiPaneInfo& paneInfo : auiMgr_.GetAllPanes())
-        paneCaptions.emplace_back(&paneInfo, paneInfo.caption);
+    wxAuiPaneInfoArray& paneArray = auiMgr_.GetAllPanes();
+    for (size_t i = 0; i < paneArray.size(); ++i)
+        paneCaptions.emplace_back(&paneArray[i], paneArray[i].caption);
 
     //compare progress dialog minimum sizes are layout-dependent + can't be changed by user => don't load stale values from config
     std::vector<std::tuple<wxAuiPaneInfo*, wxSize /*min size*/, wxSize /*best size*/>> paneConstraints;
@@ -3143,7 +3144,14 @@ void MainDialog::onSetLayoutContext(wxMouseEvent& event)
 
     bool addedSeparator = false;
 
+#if wxCHECK_VERSION(3,3,0)
     for (wxAuiPaneInfo& paneInfo : auiMgr_.GetAllPanes())
+#else
+    const wxAuiPaneInfoArray& paneArray = auiMgr_.GetAllPanes();
+    for (size_t i = 0; i < paneArray.size(); ++i)
+    {
+        wxAuiPaneInfo& paneInfo = paneArray[i];
+#endif
         if (!paneInfo.IsShown() &&
             paneInfo.window != compareStatus_->getAsWindow() &&
             paneInfo.window != m_panelLog                    &&
@@ -3161,6 +3169,9 @@ void MainDialog::onSetLayoutContext(wxMouseEvent& event)
                 this->auiMgr_.Update();
             });
         }
+#if !wxCHECK_VERSION(3,3,0)
+    }
+#endif
 
     menu.popup(*this);
 }
diff --git a/FreeFileSync/Source/ui/rename_dlg.cpp b/FreeFileSync/Source/ui/rename_dlg.cpp
index 5bd5be7..503063a 100644
--- a/FreeFileSync/Source/ui/rename_dlg.cpp
+++ b/FreeFileSync/Source/ui/rename_dlg.cpp
@@ -155,7 +155,7 @@ public:
         }
     }
 
-    int getBestSize(const wxReadOnlyDC& dc, size_t row, ColumnType colType) override
+    int getBestSize(wxDC& dc, size_t row, ColumnType colType) override
     {
         // -> synchronize renderCell() <-> getBestSize()
         return dc.GetTextExtent(getValue(row, colType)).GetWidth() + 2 * getColumnGapLeft() + dipToWxsize(1); //gap on left and right side + border
@@ -282,12 +282,13 @@ RenameDialog::RenameDialog(wxWindow* parent,
             longNames.resize(10);
         }
 
-        wxInfoDC infoDc(m_gridRenamePreview);
-        infoDc.SetFont(m_gridRenamePreview->GetFont()); //the font parameter of GetTextExtent() is not evaluated on OS X, wxWidgets 2.9.5, so apply it to the DC directly!
+        wxMemoryDC dc;
+        setScaleFactor(dc, getScreenDpiScale());
+        dc.SetFont(m_gridRenamePreview->GetFont()); //the font parameter of GetTextExtent() is not evaluated on OS X, wxWidgets 2.9.5, so apply it to the DC directly!
 
         int maxStringWidth = 0;
         for (const std::wstring& str : longNames)
-            maxStringWidth = std::max(maxStringWidth, infoDc.GetTextExtent(str).GetWidth());
+            maxStringWidth = std::max(maxStringWidth, dc.GetTextExtent(str).GetWidth());
 
         const int defaultColWidthOld = maxStringWidth + 2 * GridData::getColumnGapLeft() + dipToWxsize(1) /*border*/ + dipToWxsize(10) /*extra space: less cramped*/;
         const int defaultColWidthNew = maxStringWidth + 2 * GridData::getColumnGapLeft() + dipToWxsize(1) /*border*/ + dipToWxsize(50) /*extra space: for longer new name*/;
@@ -315,10 +316,11 @@ RenameDialog::RenameDialog(wxWindow* parent,
         m_staticlinePreview               ->Hide();
         m_staticTextPlaceholderDescription->Hide();
 
-        wxInfoDC infoDc(m_textCtrlNewName);
-        infoDc.SetFont(m_textCtrlNewName->GetFont()); //the font parameter of GetTextExtent() is not evaluated on OS X, wxWidgets 2.9.5, so apply it to the DC directly!
+        wxMemoryDC dc; //the context used for bitmaps
+        setScaleFactor(dc, getScreenDpiScale());
+        dc.SetFont(m_textCtrlNewName->GetFont()); //the font parameter of GetTextExtent() is not evaluated on OS X, wxWidgets 2.9.5, so apply it to the DC directly!
 
-        const int textCtrlDefaultWidth = std::min(infoDc.GetTextExtent(renamePhrase).GetWidth() + 20 /*borders (non-DIP!)*/ +
+        const int textCtrlDefaultWidth = std::min(dc.GetTextExtent(renamePhrase).GetWidth() + 20 /*borders (non-DIP!)*/ +
                                                   dipToWxsize(50) /*extra space: for longer new name*/,
                                                   dipToWxsize(900));
         m_textCtrlNewName->SetMinSize({textCtrlDefaultWidth, -1});
diff --git a/FreeFileSync/Source/ui/small_dlgs.cpp b/FreeFileSync/Source/ui/small_dlgs.cpp
index 343661f..e0286e8 100644
--- a/FreeFileSync/Source/ui/small_dlgs.cpp
+++ b/FreeFileSync/Source/ui/small_dlgs.cpp
@@ -1558,7 +1558,7 @@ OptionsDlg::OptionsDlg(wxWindow* parent, GlobalConfig& globalCfg) :
 
         class : public wxGridColumnHeaderRendererDefault
         {
-            void DrawHighlighted(const wxGrid& grid, wxDC& dc, wxRect& rect, int col, int flags) const override { DrawBorder(grid, dc, rect); }
+            void DrawHighlighted(const wxGrid& grid, wxDC& dc, wxRect& rect, int col, int flags) const { DrawBorder(grid, dc, rect); }
         } colRenderNoHighlight_;
     };
     m_gridCustomCommand->GetTable()->SetAttrProvider(new wxGridCellAttrProviderNoColHighlight);
diff --git a/FreeFileSync/Source/ui/tree_grid.cpp b/FreeFileSync/Source/ui/tree_grid.cpp
index 7bf9401..8b9ce4f 100644
--- a/FreeFileSync/Source/ui/tree_grid.cpp
+++ b/FreeFileSync/Source/ui/tree_grid.cpp
@@ -928,7 +928,7 @@ private:
         }
     }
 
-    int getBestSize(const wxReadOnlyDC& dc, size_t row, ColumnType colType) override
+    int getBestSize(wxDC& dc, size_t row, ColumnType colType) override
     {
         // -> synchronize renderCell() <-> getBestSize() <-> getMouseHover()
 
@@ -946,7 +946,7 @@ private:
                    2 * gapSize_; //include gap from right!
     }
 
-    HoverArea getMouseHover(const wxReadOnlyDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) override
+    HoverArea getMouseHover(wxDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) override
     {
         if (static_cast<ColumnTypeOverview>(colType) == ColumnTypeOverview::folder)
             if (std::unique_ptr<TreeView::Node> node = getDataView().getLine(row))
diff --git a/wx+/darkmode.cpp b/wx+/darkmode.cpp
index c4a3758..ebeb9a8 100644
--- a/wx+/darkmode.cpp
+++ b/wx+/darkmode.cpp
@@ -53,7 +53,8 @@ void zen::colorThemeInit(wxApp& app, ColorTheme colTheme) //throw FileError
 {
     assert(!refGlobalColorHook());
 
-    globalDefaultThemeIsDark = wxSystemSettings::GetAppearance().AreAppsDark();
+    //globalDefaultThemeIsDark = wxSystemSettings2::GetAppearance().AreAppsDark();
+    globalDefaultThemeIsDark = false;
     ZEN_ON_SCOPE_EXIT(if (!refGlobalColorHook()) refGlobalColorHook() = std::make_unique<SysColorsHook>()); //*after* SetAppearance() and despite errors
 
     //caveat: on macOS there are more themes than light/dark: https://developer.apple.com/documentation/appkit/nsappearance/name-swift.struct
@@ -86,10 +87,10 @@ void zen::changeColorTheme(ColorTheme colTheme) //throw FileError
     try
     {
         ZEN_ON_SCOPE_SUCCESS(refGlobalColorHook() = std::make_unique<SysColorsHook>()); //*after* SetAppearance()
-        if (wxApp::AppearanceResult rv = wxTheApp->SetAppearance(colTheme);
-            rv != wxApp::AppearanceResult::Ok)
-            throw SysError(formatSystemError("wxApp::SetAppearance",
-                                             rv == wxApp::AppearanceResult::CannotChange ? L"CannotChange" : L"Failure", L"" /*errorMsg*/));
+        //if (wxApp::AppearanceResult rv = wxTheApp->SetAppearance(colTheme);
+        //    rv != wxApp::AppearanceResult::Ok)
+        //    throw SysError(formatSystemError("wxApp::SetAppearance",
+        //                                     rv == wxApp::AppearanceResult::CannotChange ? L"CannotChange" : L"Failure", L"" /*errorMsg*/));
     }
     catch (const SysError& e) { throw FileError(_("Failed to update the color theme."), e.toString()); }
 }
diff --git a/wx+/darkmode.h b/wx+/darkmode.h
index 91d2a78..ad730f8 100644
--- a/wx+/darkmode.h
+++ b/wx+/darkmode.h
@@ -9,6 +9,7 @@
 
 #include <zen/file_error.h>
 #include <wx/app.h>
+#include <wx/settings.h>
 
 
 namespace zen
@@ -16,7 +17,8 @@ namespace zen
 bool darkModeAvailable();
 
 //support not only "dark mode" but dark themes in general
-using ColorTheme = wxApp::Appearance; //why reinvent the wheel?
+enum Appearance { System, Light, Dark };
+using ColorTheme = Appearance; //why reinvent the wheel?
 
 void colorThemeInit(wxApp& app, ColorTheme colTheme); //throw FileError
 void colorThemeCleanup();
@@ -25,4 +27,28 @@ bool equalAppearance(ColorTheme colTheme1, ColorTheme colTheme2);
 void changeColorTheme(ColorTheme colTheme); //throw FileError
 }
 
+struct wxColorHook
+{
+    virtual ~wxColorHook() {}
+    virtual wxColour getColor(wxSystemColour index) const = 0;
+};
+WXDLLIMPEXP_CORE inline std::unique_ptr<wxColorHook>& refGlobalColorHook()
+{
+    static std::unique_ptr<wxColorHook> globalColorHook;
+    return globalColorHook;
+}
+
+
+class WXDLLIMPEXP_CORE wxSystemSettings2 : public wxSystemSettingsNative
+{
+public:
+    static wxColour GetColour(wxSystemColour index)
+    {
+        if (refGlobalColorHook())
+			return refGlobalColorHook()->getColor(index);
+
+        return wxSystemSettingsNative::GetColour(index);
+    }
+};
+
 #endif //DARKMODE_H_754298057018
diff --git a/wx+/grid.cpp b/wx+/grid.cpp
index 54b10e5..bd2e083 100644
--- a/wx+/grid.cpp
+++ b/wx+/grid.cpp
@@ -161,7 +161,7 @@ void GridData::renderCell(wxDC& dc, const wxRect& rect, size_t row, ColumnType c
 }
 
 
-int GridData::getBestSize(const wxReadOnlyDC& dc, size_t row, ColumnType colType)
+int GridData::getBestSize(wxDC& dc, size_t row, ColumnType colType)
 {
     return dc.GetTextExtent(getValue(row, colType)).GetWidth() + 2 * getColumnGapLeft() + dipToWxsize(1); //gap on left and right side + border
 }
@@ -472,7 +472,7 @@ public:
 
     int getBestWidth(ptrdiff_t rowFrom, ptrdiff_t rowTo)
     {
-        wxInfoDC dc(this);
+        wxClientDC dc(this);
         dc.SetFont(GetFont()); //harmonize with RowLabelWin::render()!
 
         int bestWidth = 0;
@@ -1099,7 +1099,7 @@ private:
             {
                 if (0 <= row && row < rowCount && cpi.colType != ColumnType::none)
                 {
-                    wxInfoDC dc(this);
+                    wxClientDC dc(this);
                     dc.SetFont(GetFont());
                     return prov->getMouseHover(dc, row, cpi.colType, cpi.cellRelativePosX, cpi.colWidth);
                 }
@@ -1133,7 +1133,7 @@ private:
             {
                 if (0 <= row && row < rowCount && cpi.colType != ColumnType::none)
                 {
-                    wxInfoDC dc(this);
+                    wxClientDC dc(this);
                     dc.SetFont(GetFont());
                     return prov->getMouseHover(dc, row, cpi.colType, cpi.cellRelativePosX, cpi.colWidth);
                 }
@@ -1264,7 +1264,7 @@ private:
             {
                 if (0 <= row && row < rowCount && cpi.colType != ColumnType::none)
                 {
-                    wxInfoDC dc(this);
+                    wxClientDC dc(this);
                     dc.SetFont(GetFont());
                     return prov->getMouseHover(dc, row, cpi.colType, cpi.cellRelativePosX, cpi.colWidth);
                 }
@@ -2252,7 +2252,8 @@ int Grid::getBestColumnSize(size_t col) const
     {
         const ColumnType type = visibleCols_[col].type;
 
-        wxInfoDC dc(mainWin_);
+        //wxInfoDC dc(mainWin_);
+        wxClientDC dc(mainWin_);
         dc.SetFont(mainWin_->GetFont()); //harmonize with MainWin::render()
 
         const auto& [rowFirst, rowLast] = getVisibleRows(mainWin_->GetClientRect());
diff --git a/wx+/grid.h b/wx+/grid.h
index 9ee811b..5a49a97 100644
--- a/wx+/grid.h
+++ b/wx+/grid.h
@@ -107,8 +107,8 @@ public:
     virtual std::wstring getValue(size_t row, ColumnType colType) const = 0;
     virtual void         renderRowBackgound(wxDC& dc, const wxRect& rect, size_t row,                     bool enabled, bool selected, HoverArea rowHover); //default implementation
     virtual void         renderCell        (wxDC& dc, const wxRect& rect, size_t row, ColumnType colType, bool enabled, bool selected, HoverArea rowHover);
-    virtual int          getBestSize  (const wxReadOnlyDC& dc, size_t row, ColumnType colType); //must correspond to renderCell()!
-    virtual HoverArea    getMouseHover(const wxReadOnlyDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) { return HoverArea::none; }
+    virtual int          getBestSize  (wxDC& dc, size_t row, ColumnType colType); //must correspond to renderCell()!
+    virtual HoverArea    getMouseHover(wxDC& dc, size_t row, ColumnType colType, int cellRelativePosX, int cellWidth) { return HoverArea::none; }
     virtual std::wstring getToolTip                           (size_t row, ColumnType colType, HoverArea rowHover) { return std::wstring(); }
 
     //label area:
